### DISCUSSION TITLE
Swarm Dial taking forever

### DISCUSSION THREAD

**Discussion by KickedDroid** [libp2p/rust-libp2p#2579] [Q&A] [answered]
Hello, I was wondering if someone could explain as to why this code causes the swarm dial to take upwards of 30s. 
```Rust
SwarmBuilder::new(transport, behaviour, local_peer_id)
            .executor(Box::new(|fut| block_on(fut)))
            .build()
````
as opposed to this which connects to the dialed peer within ms of dialing. Note: both peer on the same local network
```Rust
Swarm::new(transport, behaviour, local_peer_id)
```

and here's where I'm dialing from 
```Rust
if matches.value_of("connect").is_some() {
        let addr: Multiaddr = matches.value_of("connect").unwrap().parse().unwrap();
        if let Err(e) = swarm.dial(addr) {
            println!("Publish error: {:?}", e);
        }
    }
````

**Comment by mxinden** [comment, marked as answer]
```
            .executor(Box::new(|fut| block_on(fut)))
```

This will block the whole thread until the `fut` is resolved. Thus nothing within `Swarm` (everything on the current thread) can make progress in the meantime. I am surprised this works at all (within 30s). I would expect it to deadlock.

See https://docs.rs/futures/latest/futures/executor/fn.block_on.html for details.

**Reply by KickedDroid** [reply]
Thanks man! Appreciated
